obspy/io/rg16/core.py

```diff
@@ -0,0 +1,236 @@
+# -*- coding: utf-8 -*-
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA
+
+import numpy as np
+from obspy.core import Stream, Trace, Stats, UTCDateTime
+
+from .util import _read, open_file, _read_block, _quick_merge
+
+# --------------------- define specs of needed blocks
+
+
+# blocks are specified as a list of tuples. Each tuple contains the following:
+# (name, startbyte, length, format), explanation as follows:
+# name - The name, these will be keys in a returned dict
+# startbye - the byte position, relative to the block, to start reading
+# length - the number of bytes to read
+# format - format to interpret the data being read. see rg16.utils.read.
+
+
+# header block, combines header block one and two
+general_header_block = [
+    ('channel_sets', 28, 1, 'bcd'),
+    ('num_additional_headers', 11, 1, '>i.'),
+    ('extended_headers', [30, 37], [1, 2], ['bcd', '>i2']),
+    ('external_headers', [31, 39], [1, 3], ['bcd', '>i3']),
+    ('record_length', 46, 3, '>i3'),
+    ('base_scan', 22, 1, '>i1'),  # https://imgur.com/a/4aneG
+]
+
+# channel set header block
+channel_header_block = [
+    ('chan_num', 1, 1, 'bcd'),
+    ('num_channels', 8, 2, 'bcd'),
+    ('ru_channel_number', 30, 1, '>i1'),
+]
+
+# combines extended header 1, 2, and 3
+extended_header_block = [
+    ('num_records', 16 + 32, 4, '>i4'),
+    ('num_files', 20 + 32, 4, '>i4'),
+    ('collection_method', 32 + 15, 1, '>i1'),
+    ('line_number', 64, 4, '>i4'),
+    ('receiver_point', 68, 4, '>i4'),
+    ('point_index', 69, 1, '>i1'),
+]
+
+# combines trace header blocks 0 (20 byte) and 1 to 10 (32 byte)
+trace_header_block = [
+    ('trace_number', 4, 2, 'bcd'),
+    ('num_ext_blocks', 9, 1, '>i1'),
+    ('line_number', 20 + 0, 3, '>i3'),
+    ('point', 20 + 3, 3, '>i3'),
+    ('index', 20 + 6, 1, '>i1'),
+    ('samples', 20 + 7, 3, '>i3'),
+    ('channel_code', 20 + 20, 1, '>i1'),  # https://imgur.com/a/4aneG
+    ('trace_count', 20 + 21, 4, '>i4'),
+    ('time', 20 + 2 * 32, 8, '>i8'),
+]
+
+# since UTCDateTime cannot be compared to np.inf in py27 get a large timestamp
+# after which I will be dead (somebody else's problem)
+BIG_TS = UTCDateTime('3000-01-01').timestamp
+
+# default channel mapping
+COMPONENT_MAP = {2: 'Z', 3: '1', 4: '2'}
+
+# map sampling rate to band code according to seed standard
+BAND_MAP = {2000: 'G', 1000: 'G', 500: 'D', 250: 'D'}
```

Comment:
we have something similar already, maybe it's a good idea to put it there in `core.util`.. a helper function to get appropriate band code for sampling rate and broadband yes/no..

see https://github.com/obspy/obspy/blob/5762fb5097404dbb04bf91fba6ecacb1075f0bdf/obspy/core/util/misc.py#L42 and https://github.com/obspy/obspy/blob/5762fb5097404dbb04bf91fba6ecacb1075f0bdf/obspy/core/util/misc.py#L66

Reply:
Looking closer at this I don't think it would be a great idea to do this in the read_rg16 function, primarily for efficiency reasons. The rg16 files can be quite large (sometimes 10 GB or more) with thousands to hundreds of thousands of traces. Reading these files is already slower than I would like. 

Since the nodes can only have sampling_rates  of 2000, 1000, 500, and 250 getting the band code is a simple dictionary lookup as it stands now. If I created a general function, it would incur the cost of a function call, as well as bounds checking to figure out which bin the sampling rate falls into, then additional cost for selecting between instruments of different corner period.

Reply:
I could be over optimizing and the cost might be negligible, I am not sure.